[CORE-8949] Incorporate datalake into space management accounting

dotnwat opened the pull request:
* Replaces https://github.com/redpanda-data/redpanda/pull/25048
* Adds account of disk usage by datalake in space management

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat committed:
datalake: use a configurable scratch space

Prior to this change the system temporary directory was used. Instead,
we should use a scratch space under the configured data directory so
that we can keep all redpanda data in one place and avoid accidental
problems like filling up tmpfs or other things.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
storage: account for datalake scratch space usage

The space used by datalake is not reclaimable, but we still want space
management to track it so it has an accurate view of space usage.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

bharathv commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -370,4 +371,40 @@ void datalake_manager::stop_translator(const model::ntp& ntp) {
+    }
+
+    chunked_vector<std::filesystem::path> files;
+    co_await directory_walker::walk(
+      path.string(), [&files, path](const ss::directory_entry& de) {
+          if (de.type == ss::directory_entry_type::regular) {
+              files.push_back(path / std::filesystem::path(de.name));
+          }
+          return ss::now();
+      });
```
curious is there no utility to compute the size of a directory? Do we always have to walk the directory  and compute sizes of individual files?

dotnwat replied:
file systems generally don't provide roll-up directory statistics (and in particular, I don't believe XFS does). so the only solutions are application-specific. what you can do here (and what we do for storage layer) is attempt to cache the size, which works well for segments that get created and then exist unchanged for long periods of time. for this scratch space files, there is probably lots of churn so it's not clear that the caching approach will have any benefit.

bharathv commented on src/v/resource_mgmt/storage.h:
```diff
@@ -13,6 +13,7 @@
 
 #include "base/seastarx.h"
 #include "config/property.h"
+#include "datalake/datalake_manager.h"
```
this include can go in the source file?

dotnwat replied:
nice catch, fixed

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1c56f3a5-046d-4d63-b2e7-06cf3dba6b6c-->
<!--TEST RESULTS-->
<!--52e2f1de-78bb-4471-b55c-a5beb8c01a79-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61652">build#61652</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61652#0194da14-492b-4a76-80d8-37495e287130 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61652#0194da14-492c-4418-bac0-a088e929e1a4 | FLAKY | 1/3 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=True.server_cfg=300000.topic_cfg=None | ducktape | https://buildkite.com/redpanda/redpanda/builds/61652#0194d9f8-622a-40f4-a4d9-ea6c1af0536f | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/61652#0194da14-492b-4a76-80d8-37495e287130 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61652#0194da14-492d-4bd8-a770-7881c0273c94 | FLAKY | 1/2 |
| rptest.tests.timequery_test.TimeQueryKafkaTest.test_timequery_below_start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/61652#0194d9f8-6229-4371-80e0-e78c1f06592a | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61700">build#61700</a></summary>


| test_id                                                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61700#0194ddd4-425e-4cc4-b2f6-bf50e21c8fe0 | FLAKY         | 1/3      |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61700#0194ddd4-425b-4d60-ba4d-36de12fbea21 | FLAKY         | 1/2      |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61700#0194ddd4-425b-4d60-ba4d-36de12fbea21 | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=2.cloud_storage_type=CloudStorageType.ABS                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61700#0194ddd4-425b-4d60-ba4d-36de12fbea21 | FLAKY         | 1/2      |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61700#0194ddba-1725-42d6-bc9e-8962da5e52a3 | FLAKY         | 1/2      |
</details>
<!--/body-->

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -370,4 +371,40 @@ void datalake_manager::stop_translator(const model::ntp& ntp) {
+            .then([&total](uint64_t size) { total += size; })
+            .handle_exception([path](std::exception_ptr eptr) {
+                vlog(
+                  datalake_log.debug,
```
nit: would it be too noisy to log this at warn level? Seems like it could be helpful in an incident to see right away if it's causing space management to behave badly

dotnwat replied:
we could log errors at warn level. except that it is normal for the occasional file_does_not_exist error to occur (file listing, stating, and removal all race), so we'd probably want to pluck that one out and either log it at INFO or DEBUG. wdyt?

andrwng replied:
Ah yea good call, file_does_not_exist at debug sounds good

dotnwat replied:
fixed

dotnwat committed:
datalake: log fs errors at warn except ENOENT

ENOENT is not unexpected since file listing and stat will race with file
removal.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>
